tracks/python/exercises/palindrome-products/mentoring.md

```diff
@@ -0,0 +1,172 @@
+
+## Faster solution
+
+```python
+from typing import Iterator, Optional
+
+Factors = list[list[int]]
+Result = tuple[Optional[int], Factors]
+
+
+def palindromes(digits: int, *, reverse: bool = False,
+                add_zero: bool = False) -> Iterator[int]:
+    # Palindromes are:
+    # - [digit]: 1, 2, ..., 9
+    # - [digit][same_digit]: 11, 22, ..., 99
+    # - [digit][palindrome][same_digit]: e.g. 4(56765)4
+    if reverse:
+        outer_range = range(9, -1 if add_zero else 0, -1)
+    else:
+        outer_range = range(0 if add_zero else 1, 10)
+    for outer in outer_range:
+        if digits == 1:
+            yield outer
+        elif digits == 2:
+            yield outer * 11
+        else:
+            for inner in palindromes(digits-2
+                                     reverse=reverse,
+                                     add_zero=True):
+                number = outer*10**(digits-1) + inner*10 + outer
+                yield number
+
+
+def get_factors(number: int, min_factor: int, max_factor: int) -> Factors:
+    factors: Factors = []
+    # number = i*j
+    # i in [1, sqrt(number)] when i <= j
+    # i, j in [min_factor, max_factor]
+    # i, j in [p//max_factor, p//min_factor]
+    for i in range(max(min_factor, number//max_factor),
+                   min(int(number**0.5), max_factor,
+                       number//min_factor) + 1):
+        if number % i == 0:
+            j = number // i
+            if j < i:
+                break
+            if j >= min_factor and j <= max_factor:
+                factors.append([i, j])
+    return factors
+
+
+def search(min_factor: int, max_factor: int, *,
+           reverse: bool = False) -> Result:
+    min_p, max_p = (min_factor*min_factor, max_factor*max_factor)
+    min_digits, max_digits = (len(str(min_p)), len(str(max_p)))
+    if reverse:
+        digits_range = range(max_digits, min_digits-1, -1)
+    else:
+        digits_range = range(min_digits, max_digits)
+    for digits in digits_range:
+        for p in palindromes(digits, reverse=reverse):
+            if p >= min_p and p <= max_p:
+                factors = get_factors(p, min_factor, max_factor)
+                if factors:
+                    return p, factors
+    return None, []
+
+
+def largest(min_factor: int, max_factor: int) -> Result:
+    if min_factor > max_factor:
+        raise ValueError('min must be <= max')
+    return search(min_factor, max_factor, reverse=True)
+
+
+def smallest(min_factor: int, max_factor: int) -> Result:
+    if min_factor > max_factor:
+        raise ValueError('min must be <= max')
+    return search(min_factor, max_factor)
+```
+
+## Performance comparison
+
+Let's say ***N*** is the size of the factor range and ***f*** is the frequency of palindromes within integer domain. The latter is closer to 0 than 1. The naive solution will have an average time complexity of **O(N²)**, the reasonable solution will have **O(N²f)** and the faster solution will have **O(Nf)**.
```

Comment:
This line also deviates from the one sentence, one line, style that has been present throughout this document (and is the style we are adopting).

Reply:
Updated the flow of this section.

The information density is still high, but this can be useful if the discussion goes into time complexity. Let me know if you think I should make it even more brief.

Reply:
These are mentor notes, so I think those can be a bit dense, but you are
likely to get more feedback from mentors than from students.  I am
saying that letting that go here, we will be able to iterate it on the
longer term.